controller/golden_files_test.go

```diff
@@ -0,0 +1,121 @@
+// -update flag is given, that golden file is overwritten with the current
+// actual object. When adding new tests you first must run them with the -update
+// flag in order to create an initial golden version.
+func compareWithGolden(t *testing.T, goldenFile string, actualObj interface{}) {
+	err := testableCompareWithGolden(*updateGoldenFiles, goldenFile, actualObj)
+	if err != nil {
+		t.Fatalf("%+v", err)
+	}
+}
+
+func testableCompareWithGolden(update bool, goldenFile string, actualObj interface{}) error {
+	absPath, err := filepath.Abs(goldenFile)
+	if err != nil {
+		return errs.WithStack(err)
+	}
+	actual, err := json.MarshalIndent(actualObj, "", "  ")
+	if err != nil {
+		return errs.WithStack(err)
+	}
+	if update {
+		err = ioutil.WriteFile(absPath, actual, os.ModePerm)
+		if err != nil {
+			return errs.Wrapf(err, "failed to update golden file: %s", absPath)
+		}
+	}
+	expected, err := ioutil.ReadFile(absPath)
+	if err != nil {
+		return errs.Wrapf(err, "failed to read golden file: %s", absPath)
+	}
+
+	expectedStr := string(expected)
+	actualStr := string(actual)
+	if expectedStr != actualStr {
+		dmp := diffmatchpatch.New()
+		diffs := dmp.DiffMain(expectedStr, actualStr, false)
+		return errs.Errorf("mismatch of actual output and golden-file %s:\n %s \n", absPath, dmp.DiffPrettyText(diffs))
+	}
+	return nil
+}
+
+func TestCompareWithGolden(t *testing.T) {
+	resource.Require(t, resource.UnitTest)
+	t.Parallel()
+	type Foo struct {
+		Bar string
+	}
+	dummy := Foo{Bar: "hello world"}
+	t.Run("file not found", func(t *testing.T) {
+		t.Parallel()
+		// given
+		f := "not_existing_file.golden.json"
+		// when
+		err := testableCompareWithGolden(false, f, dummy)
+		// then
+		require.NotNil(t, err)
+		_, isPathError := errs.Cause(err).(*os.PathError)
+		require.True(t, isPathError)
+	})
+	t.Run("unable to update golden file due to not existing folder", func(t *testing.T) {
+		t.Parallel()
+		// given
+		f := "not/existing/folder/file.golden.json"
+		// when
+		err := testableCompareWithGolden(true, f, dummy)
+		// then
+		require.NotNil(t, err)
+		_, isPathError := errs.Cause(err).(*os.PathError)
+		require.True(t, isPathError)
+	})
+	t.Run("mismatch between expected and actual output", func(t *testing.T) {
+		t.Parallel()
+		// given
+		f := "test-files/codebase/show/ok_without_auth.golden.json"
+		// when
+		err := testableCompareWithGolden(false, f, dummy)
+		// then
+		require.NotNil(t, err)
+		_, isPathError := errs.Cause(err).(*os.PathError)
+		require.False(t, isPathError)
+	})
+	t.Run("ok - expected output equals actual", func(t *testing.T) {
+		t.Parallel()
+		// given
+		f := "test-files/dummy.golden.json"
+		bs, err := json.MarshalIndent(dummy, "", "  ")
+		require.Nil(t, err)
+		err = ioutil.WriteFile(f, bs, os.ModePerm)
+		require.Nil(t, err)
+		// when
+		err = testableCompareWithGolden(false, f, dummy)
+		// then
+		assert.Nil(t, err)
+		err = os.Remove(f)
```

Comment:
@alexeykazakov fixed in 3b551fb. To be honest I've thought about using defer but when I just wrote

```go
defer require.Nil(t, os.Remove(f))
```

the file was deleted immediately. I must have left it in the code as a normal call to have it removed at all.

Thanks for finding this left-over.